### PR TITLE
support deserializeState

### PR DESCRIPTION
**Pull request by baryon**

### CODE REVIEW

File: src/serializer.ts [hunk 1]
```diff
@@ -1,63 +1,125 @@
-import { bsv, num2bin } from './utils';
+import { bin2num, bsv, num2bin } from './utils'
 
 /*
  * a varint serializer into Script ASM
  */
-const Script = bsv.Script;
-const BN = bsv.crypto.BN;
+const Script = bsv.Script
```
Comment (xhliu) [inline src/serializer.ts:6]: Let's do not change format. Or put formatting in a separate commit if needed.

File: src/serializer.ts [hunk 2]
```diff
@@ -1,63 +1,125 @@
...
+    return this.op.buf.toString(arg);
+  }
+}
+
+export type OpStateArray = Array<OpState>
+
+// deserialize Script or Script Hex to contract state array
+// DONT support ASM Code converting.
+// TODO: validate
+export function deserializeState(s: string | bsv.Script): OpStateArray {
```
Comment (xhliu) [inline src/serializer.ts:110]: Hmm, this is different from what I thought:
```
deserializeState<T>(s: string | bsv.Script): T

StateClass s = { name: "Bob", age: 5}
// desearize is the opposite of serialize, so should return the same object
expect(deserializeState<StateClass>(serializeState(s)) == s)
```

**Review by xhliu — CHANGES_REQUESTED** [review]
See if you can deserialize using [Generics](https://www.typescriptlang.org/docs/handbook/generics.html).

File: src/serializer.ts (hunk 2 above)
Comment (baryon) [reply src/serializer.ts:110]: do it, no problem

**Review by xhliu — APPROVED** [review]
